futures-util/src/future/poll_fn.rs

```diff
@@ -34,18 +36,18 @@ pub struct PollFn<F> {
 /// # }
 /// ```
 pub fn poll_fn<T, E, F>(f: F) -> PollFn<F>
-    where F: FnMut(&mut task::Context) -> Poll<T, E>
+    where F: FnMut(&mut task::Context) -> Poll<T>
 {
     PollFn { inner: f }
 }
 
-impl<T, E, F> Future for PollFn<F>
-    where F: FnMut(&mut task::Context) -> Poll<T, E>
+impl<T, F> Future for PollFn<F>
+    where F: FnMut(&mut task::Context) -> Poll<T>
 {
-    type Item = T;
-    type Error = E;
+    type Output = T;
 
-    fn poll(&mut self, cx: &mut task::Context) -> Poll<T, E> {
-        (self.inner)(cx)
+    fn poll(mut self: Pin<Self>, cx: &mut task::Context) -> Poll<T> {
+        // safe because we never expose a Pin<F>
+        (unsafe { Pin::get_mut(&mut self) }.inner)(cx)
```

Comment:
There's a few places where you get mutable references to `Unpin` fields (well, in this case a field that could be `Unpin` without affecting it since `FnMut` already implies it must be `Unpin`). I have been using an additional macro when doing this to reduce the number of `unsafe` patterns needed to check:

```rust
macro_rules! unpinned_field {
    ($pin:expr, $field:ident) => {
        &mut *pinned_field!($pin, $field)
    };
}
```

(although, I'm still also using a "safe" `pinned_field` macro that relies on me not doing anything stupid in `Drop`, so I'm not sure how much benefit this gives when your `pinned_field` macro is unsafe to call).